draft-proxied-svcb-headers.md

```diff
@@ -17,132 +17,551 @@ author:
+representation of the IPv6 or IPv4 address that represents the
+destination of the CONNECT* from the proxy.  If this is an IPv6
+address, it MUST follow the canonical string form from {{!RFC5952}}.
+
+*TODO*: Is there an IPv4 canonical form RFC we should reference here?
+
+If the hostname connected to by the proxy was a series of CNAMEs,
+these should be included at the front of the list, starting with
+the CNAME that the hostname resolved to and proceeding sequentially.
+
+Each item in the list has the following parameters:
+
+* t: The DNS RR type of the DNS query name, which MUST be specified
+  as an sf-integer (so 1 for A, 28 for AAAA, and 5 for CNAME).
+
+* ttl: The TTL of the address or CNAME record SHOULD be included as a
+  parameter with the key "ttl", whose value MUST be an sf-integer.
+  If the value of of intermediate records are unavailable,
+  the ttl parameter SHOULD be included on just the last list
+  entry, containing the minimum TTL value across the CNAMEs.
+
+* o: The owner name of the DNS record MAY be included.
+  This this parameter is specified, its value MUST be an sf-string
+  whose value is the owner name of the record.
+
+
+
+### Example Proxy-DNS-Used response
+
+As an example, assume that the proxy server resolved the
+authority hostname of "svc.example.com" as follows:
 
 ~~~ diagram
-   svc.example.com. 3600 IN HTTPS 1 svc2.example.com. alpn=h2,h3 echconfig="123..."
-   svc.example.com. 3600 IN HTTPS 2 . alpn=h2 echconfig="abc..."
+   svc.example.com.  7200 IN CNAME  svc.example.net.
+   svc.example.net.  1800 IN CNAME  svc2.example.net.
+   svc2.example.net.   60 IN A      192.0.2.74
+   svc2.example.net.   60 IN AAAA   2001:db8::75
+   svc2.example.net.   60 IN AAAA   2001:db8::76
 ~~~
 
-A successful CONNECT response would include the following headers, if the client requested both
-"alpn" and "echconfig":
+If the proxy connected to 2001:db8::75 then the successful CONNECT
+response would include the following header:
 
 ~~~ example
 HEADERS
 :method = CONNECT
 :status = 200
-dns-svcb-params = "svc2.example.com.";priority=1;ttl=3600;p1=:aDIsaDM=:;p5=:MTIzLi4u:,
-                  "svc.example.com.";priority=2;ttl=3600;p1=:aDI=:;p5=:YWJjLi4u:
+proxy-dns-used = "svc.example.net.";ttl=7200;t=5;o="svc.example.com.",
+                 "svc2.example.net.";ttl=1800;t=5;o="svc.example.net.",
+                 "2001:db8::75";ttl=60;t=28;o="svc2.example.net."
 ~~~
 
+
+*TODO*: Do we need the "o=" parameter or is it redundant?
+        If we include it, should it be a MUST or MAY?
+        
+*TODO*: Do we need to cover DNAME as well? 
+
+*TODO*: Should a future version be able to include NS record
+        and DNSSEC information?
+
+
+# Proxy Behavior
+
+Proxy servers MUST NOT take action based on SVCB records.
+In particular, the ipv4hint and ipv6hint SvcParams MUST NOT
+be used by proxies for making connections.
+
+*TODO*: Discuss if this is too restrictive
```

Comment:
I think this probably is too restrictive, especially in the case of SVCB-naive clients.  It means a high probability of the client having to open a new tunnel after processing the records, even though avoiding that is something this spec states as a goal.

It still seems reasonable for the proxy to chase AliasMode records, use hints, etc. to choose the target to connect to.  The flip side, I suppose, would be if the proxy chooses a host whose ALPN list doesn't reflect the client's support, but that could be solved by requiring the server to choose based on the transport protocol implied by the method, and then preferring the default ALPN within that.  Alternatively, the client could include an ALPN list in the request.

However, it's also totally reasonable to save that for a separate PR if we decide to do it.

Reply:
Part of why I'm wary of this is that it changes the Proxy behavior and semantics substantially.  Rather than the proxy just returning additional information, here we start fundamentally changing behavior.  (Would doing this count as an "Update" to any of the RFCs defining CONNECT, for example?)